Gap filler model

ecunha1996 opened the pull request:
To check the results, please unzip the data.zip file that is in the tests folder first. (the data folder was to big to push).

What I changed:

-Add a method (test_e_precursors) to check if bioiso was giving false positives (model class).

-Get all completions  instead of just the minimal (up to a maximum of 50 - can be changed in the submission_parameters.json).

-Add additional seeds (add seeds for cofactors associated with target's pathways) - check "identify_additional_seeds"

-Evaluate all completions with fba - description in the "add_reactions_to_model" method.

beach-fossils committed:
[ADD] gap filler and model class

beach-fossils committed:
[ADD] write_metabolites_to_sbml function file

beach-fossils committed:
[ADD] new docstring

beach-fossils committed:
[ADD] isolated file for Model class

beach-fossils committed:
[ADD] file for GapFiller class, from_folder method added

beach-fossils committed:
[ADD] file for GapFiller class, from_folder method

beach-fossils committed:
[ADD] file for Model class
§changed idenfity_targets method to raise errors when parameters are None

beach-fossils committed:
[ADD] Unit tests for Model class

Added a test file for the Model class, including test methods for initializing the model and identifying seeds and targets. All tests pass successfully

beach-fossils committed:
[ADD] Unit test for the write_metabolites_to_sbml function

Added a test file for the mentioned function. It passed the tests successfully

beach-fossils committed:
[UPDATE] Minor documentation

beach-fossils committed:
[ADD] init file

beach-fossils committed:
[UPDATE] Documentation at the to_sbml method

beach-fossils committed:
[ADD] Function to transform "__" into "_"

beach-fossils committed:
[ADD] Test the function to transform "__" into "_"

It passed successfully the test

beach-fossils committed:
[ADD] File to test the GapFiller class

It did pass successfully some tests. However, it was not true to:

✺ test_run: accused an error when comparing the return dict values, "'Unproducible targets': <ANY>} != None"

✺ test_evalute_results: tried to build the function, but it was giving many errors; I will try to reformulate the problem

beach-fossils committed:
[UPDATE] minor changes

beach-fossils committed:
[UPDATE] minor changes at the to_sbml method

now if there is already a file_name and save_path coincident it adds an extra number to the file_name to avoid overwriting previous files

beach-fossils committed:
[UPDATE] minor testing changes at the to_sbml method

it passed the test successfuly

beach-fossils committed:
[UPDATE] All tests passed successfully

beach-fossils committed:
[UPDATE] minor change

beach-fossils committed:
[UPDATE] minor change

beach-fossils committed:
[UPDATE] minor change at the test_to_sbml function

beach-fossils committed:
[UPDATE] alterações feitas na reunião

beach-fossils committed:
[ADD] notebook para Emanuel checkar

beach-fossils committed:
[ADD] notebook para Emanuel checkar

beach-fossils committed:
[ADD] model, seeds, targets and universal model

beach-fossils committed:
[UPDATE] add_reactions_to_model and compare_reactions methods added to the GapFIller class

beach-fossils committed:
[UPDATE] minor things

beach-fossils committed:
[ADD] test_add_reactions_to_model, test_compare_reactions and test_compare_reactions_no_added_reactions tests added

✺ test_add_reactions_to_model fails: "Infeasible" error
✺ test_compare_reactions successfully passed the tests
✺ test_compare_reactions_no_added_reactions passed the tests

beach-fossils committed:
reuniao 19mar, new_test_toy

beach-fossils committed:
testar os metodos com modelo 6gaps;
erro na última célula ao correr o metodo 'add_reactions_to_model' com o cobra.gapfilling

ecunha1996 committed:
[ADD] test for cobra gapfill;

beach-fossils committed:
[ADD] function to replace 'extr' and 'cytop' terms

This commit adds a new function, `fix_terms`, to the codebase. The function takes two arguments, an input file path and an output file path, and replaces all occurrences of the terms 'extr' and 'cytop' in the input file with 'out' and 'in', respectively. The modified data is then written to the output file.

This feature will be useful to process .XML files containing metabolic models

beach-fossils committed:
[ADD] unit test for 'fix_terms' fucntion

 The test creates a temporary input file containing the string 'extracellular cytoplasmic', calls the function to modify the terms, and checks if the output file contains the correct replacements. The test is designed to ensure that the function works correctly

beach-fossils committed:
[UPDATE] minor changes; ready to run the with the 6_gaps model

beach-fossils committed:
[UPDATE] minor change

beach-fossils committed:
[ADD] performance test for meneco; ready to run with the 6_gaps model

beach-fossils committed:
[ADD] 6_gaps model

beach-fossils committed:
[UPDATE] add_reactions_to_model and compare_reactions modified; now they should be working fine

✺ tests for those mentioned methods are still missing

beach-fossils committed:
[UPDATE] changes at test_add_reactions_to_model, test_compare_reactions and test_compare_reactions_no_added_reactions

✺ test_add_reactions_to_model takes a long time to run and won't pass the test due to Infeasible Cobra error

✺ test_compare_reactions and test_compare_reactions_no_added_reactions passed the test successfully

ecunha1996 committed:
[ADD] restrict search space by pathway

beach-fossils committed:
[UPDATE] Addition of parameter temporary_universal_model to the 'from_folder' method

This commit adds a new parameter temporary_universal_model to the from_folder method of the GapFiller class. If this parameter is set as True, a temporary universal model will be created that contains only the pathways related to the targets. This is useful for reducing the size of the universal model, improving the time taken when applying gap-filling algorithms.

beach-fossils committed:
[ASS] test for 'from_folder' method

This commit adds a test case for the GapFiller.from_folder() method with the temporary_universal_model parameter set to True. The test checks the correct creation of a temporary universal model and validates the presence of the expected model file in the specified folder. The test also ensures that the universal model is not None after the GapFiller object is created.

beach-fossils committed:
[ADD] test for 'from_folder' method
This commit adds a test case for the GapFiller.from_folder() method with the temporary_universal_model parameter set to True. The test checks the correct creation of a temporary universal model and validates the presence of the expected model file in the specified folder. The test also ensures that the universal model is not None after the GapFiller object is created.

beach-fossils committed:
[UPDATE] minor change at the 'from_folder' method

It had a bug, which was fixed during testing sessions

beach-fossils committed:
[UPDATE] [UPDATE] Reformulate tests

Reformulated the tests for better coverage and clarity after some important changes. All tests passed successfully.

beach-fossils committed:
[ADD] create_random_knockout_models method

 Method to generate multiple knockout models

beach-fossils committed:
[UPDATE] minor change at documentation

beach-fossils committed:
[ADD] 6 different models to proceed with the gap-filling analysis process

beach-fossils committed:
[ADD] Performance test methods for running multiple models within folders

beach-fossils committed:
[ADD] Knockout models data added; including: seeds, targets and temporary universal models for each knockout model

beach-fossils committed:
[ADD] method to save meneco_output as a txt file, reporting the process of gap-filling

beach-fossils committed:
[UPDATE] removed reactions txt file from each model

beach-fossils committed:
[UPDATE] create_random_knockout_models: now it checks if seeds and targets are not empty and also prints the removed reactions

beach-fossils committed:
[UPDATE] minor change

beach-fossils committed:
Delete model_spneumoniaeR6_knockout_5.xml

ecunha1996 committed:
[ADD] KEGG retrieve associated pathways

ecunha1996 committed:
Merge branch 'gap_filler_model' of github.com:jcapels/gap_filling_dl into gap_filler_model

ecunha1996 committed:
[ADD] retrieve associated pathways from KEGG

beach-fossils committed:
[ADD] notebook for the minimal completion method

beach-fossils committed:
[REFINEMENT] Incorporate minimal_completion from meneco in the 'run' method, introducing significant modifications to the display of the report

beach-fossils committed:
[ADD] Reports for the outcomes of the gap-filling minimal_completion method of meneco

beach-fossils committed:
[ADD] Initial attempt to perform gap filling on the Synechocystis model

beach-fossils committed:
[ADD] Initial attempt to perform gap filling on the Synechocystis model

beach-fossils committed:
[ADD] Initial attempt to perform gap filling on the Synechocystis model

beach-fossils committed:
Remove /Users/josediogomoura/gap_filling_dl/tests/data/models/synechocystis.xml
 from version control

ecunha1996 committed:
[ADD] retrieve associated pathways from KEGG

ecunha1996 committed:
Merge branch 'gap_filler_model' of github.com:jcapels/gap_filling_dl into gap_filler_model

beach-fossils committed:
[ADD] Various methods for the GapFiller class

1. `check_reactions`: This method verifies if a list of reactions is present in the universal model and prints an error message for those that aren't; this one is not very relevant

2. `clone_reactions`: This method clones reactions from a universal model, adjusting the compartment of the cloned metabolites. The reactions are then added to a new model that is a copy of the original one. Transport reactions are also cloned and added to the new model. The new model is stored in `self.cloned_model` and returned.

3. `get_transport_reactions`: This method retrieves and clones all the transport reactions (reactions involving metabolites in more than one compartment) from the original model.

4. `identify_dead_end_metabolites`: This method identifies dead-end metabolites in the provided model (or the original model if none is provided). If the original model is used, the result is saved to `self.dead_ends`.

Note: The above methods have been implemented but not yet tested

beach-fossils committed:
[ADD] Utility functions for compartment retrieval and dead-end metabolite identification

1. `get_compartments_in_model`: This function takes a model as input and returns a list of compartments present in the model. It iterates over the reactions in the model and extracts the compartment information from the reaction IDs.

2. `identify_dead_end_metabolites`: This function identifies dead-end metabolites in a model by checking if they participate in only one reaction. It performs flux variability analysis on these reactions and determines if the maximum and minimum flux values are both zero. The dead-end metabolites are stored in a list and returned.

Note: These functions have been implemented but not yet tested.

beach-fossils committed:
[ADD] methods to handle dead-end metabolites and clone reactions

`clone_reactions`: This method has been updated to clone reactions from the universal model into all compartments that exist in the original model. This is done using multiprocessing to speed up the process. (trying the multiprocessing)

`get_transport_reactions`: This method retrieves all transport reactions in the original model and clones them.

`get_dead_end_metabolites`: This method identifies dead-end metabolites in the model.

`identify_dead_end_metabolites_with_bioiso`: This method identifies dead-end metabolites in the model using BioISO.

`fill_dead_ends`: This method fills the dead-end metabolites in the model using transport reactions.

`get_plausible_transports_for_dead_ends`: This method creates a dictionary of plausible transport reactions for each dead-end metabolite. UNDONE

`create_and_add_transport_reaction`: This method creates a new transport reaction for a given metabolite and adds it to the cloned model. UNDONE

`is_plausible_transport`: This method checks if a given transport reaction is plausible. UNDONE

ecunha1996 committed:
[FIX] running draft model

ecunha1996 committed:
Merge branch 'gap_filler_model' of github.com:jcapels/gap_filling_dl into gap_filler_model

# Conflicts:
#	src/gap_filling_dl/biomeneco/gapfiller.py

beach-fossils committed:
[UPDATE] Optimized many methods:

Optimized the get_transport_reactions_of_draft and get_transport_reactions_of_universal methods by removing the unnecessary list comprehension and directly returning the list of transport reactions.
Optimized the identify_dead_end_metabolites_with_bioiso method by removing unnecessary checks for string types and values, and by directly accessing dictionary values instead of using the get method.
Optimized the fill_dead_ends method by removing the unnecessary check for the existence of self.cloned_model and by directly accessing self.dead_ends instead of calling the get_dead_end_metabolites method.
Optimized the add_known_transport_reactions method by removing the unnecessary check for the existence of self.transport_reactions and by directly accessing self.transport_reactions instead of calling the get_transport_reactions_of_universal method.
Optimized the add_reactions_to_model method by removing the unnecessary replacement of 'R_' in reaction IDs and by directly adding reactions to the model without checking if they are already present.

All of these havent been tested yet

beach-fossils committed:
[ADD] Creating seeds and targets file for the synechocystis model, trying to run the different methods to fill the dead-end metabolites and cloning the model

ecunha1996 committed:
[FIX] clone compartments & universal_model.xml with irreversible reactions

ecunha1996 committed:
Merge branch 'gap_filler_model' of github.com:jcapels/gap_filling_dl into gap_filler_model

# Conflicts:
#	src/gap_filling_dl/biomeneco/gapfiller.py
#	tests/performance_tests/synechocystis.ipynb

beach-fossils committed:
[ADD] Several new methods were added and from_folder method was updated

Updated method:
from_folder, cloned the model and set it as the universal_model_compartmentalized attribute if the compartmentalized file was found and the clone flag is True.

Added method: met_is_being_consumed_or_produced_at_reaction to check if a metabolite is being consumed or produced at a reaction.

Added method: check_metabolite_in_reaction to check if a metabolite is being consumed or produced at a reaction.
Iterated over the reactions and stored the involvement of the metabolite in each reaction.
Added print statements to display reaction information.

Added method:
search_dead_end to search for reactions in the universal model where a given metabolite is consumed or produced.
Implemented the functionality to search for consumed and/or produced reactions based on the provided parameters.
Added an optional verbose flag to print the reactions where the metabolite is consumed or produced.
Returned a tuple containing lists of consumed and produced reactions.

Added method:
 deal_dead_ends to handle dead-end metabolites.
Implemented the functionality to search for reactions in the universal model where the dead-end metabolite is consumed or produced.
Added logic to add the identified reactions to the draft model.
Handled the case where the dead-end metabolite is in different compartments by creating a transport reaction.
Updated the draft model with the added reactions.
Implemented the functionality to automatically identify dead-end metabolites if not provided.
Returned the updated draft model.

beach-fossils committed:
[ADD] notebook  for performing gap filling of the synechocystis model

Created a GapFiller object from the provided folder using the from_folder method.
Identified and dealt with dead-end metabolites using BioIso and the deal_dead_ends method.
Created a new Model object from the updated model and performed optimization for biomass production.
Saved the updated model to an SBML file with targets and seeds.
Created a new GapFiller object from the folder of the updated model and executed gap filling again

ecunha1996 committed:
[FIX] clone compartments & universal_model.xml with irreversible reactions

ecunha1996 committed:
Merge branch 'gap_filler_model' of github.com:jcapels/gap_filling_dl into gap_filler_model

# Conflicts:
#	src/gap_filling_dl/biomeneco/gapfiller.py

beach-fossils committed:
[ADD] 'add_transport_reaction_for_dead_ends' method

The method handles dead-end metabolites by creating transport reactions between different compartments to facilitate metabolite movement between the compartments. It is called at 'from_folder' method. Not tested yet.

beach-fossils committed:
[FIX] Replaced ValueError with a warning for scenarios with less than two compartments; ensures continued execution.

beach-fossils committed:
[UDD] Dockerfile,service_main.py and requirements.txt update

beach-fossils committed:
Remove tests/data/draft_synechocystis/ from git tracking

beach-fossils committed:
Update .gitignore to exclude tests/data/draft_synechocystis/

ecunha1996 committed:
[FIX] merge

ecunha1996 committed:
Merge branch 'gap_filler_model' of github.com:jcapels/gap_filling_dl into gap_filler_model

# Conflicts:
#	requirements.txt
#	tests/data/draft_synechocystis/draft_synechocystis_seeds.xml
#	tests/data/draft_synechocystis/draft_synechocystis_targets.xml
#	tests/data/draft_synechocystis/model_synechocystis.xml
#	tests/data/draft_synechocystis/universal_model.xml

ecunha1996 committed:
[FIX] merge & solve bioiso installation

ecunha1996 committed:
[FIX] Dockerfile  ADD cplex

ecunha1996 committed:
[FIX] Dockerfile  ADD cplex

ecunha1996 committed:
[FIX] Dockerfile & remove possible metabolites duplication

beach-fossils committed:
[UPDATE] report method is now structured as a dict, and the final file format as a JSON

beach-fossils committed:
[UPDATE] get_metabolite_pathway_map at utils.py was changed for debugging. Several print statements were added at gapfiller.py. MenecoRunner.py was changed to run the program locally. model_1 added for testing purposes

beach-fossils committed:
[ADD] minimal completion section added at the report method

ecunha1996 committed:
[FIX] gapfiller with results && nice progress bar

beach-fossils committed:
[ADD] Example JSON file required for the workflow

ecunha1996 committed:
[FIX] add additional demands  get more solutions

ecunha1996 committed:
[FIX] add additional demands  get more solutions

ecunha1996 committed:
Merge branch 'gap_filler_model' of https://github.com/jcapels/gap_filling_dl into gap_filler_model

ecunha1996 committed:
[FIX] add additional demands  get more solutions

ecunha1996 committed:
[FIX] compressed data folder

jcapels commented on .DS_Store:
this file has to be removed from github and added to the gitignore file

jcapels commented on examples/meneco_with_pathways.py:
```diff
@@ -0,0 +1,197 @@
+            #     print(f"Error for {model_name}: {type(e).__name__}: {str(e)}")
+            #     continue
+
+        # Save results to a file after all models have been processed
+        self.save_results_to_file()
+
+    def print_gapfilling_stats(self, model_name, results, file=None):
+        # print_or_write("Reactions removed dictionary keys: " + ", ".join(self.reactions_removed.keys()), file)
+
+        # if model_name not in self.reactions_removed:
+        #     print_or_write(f"{model_name} not found in reactions_removed dictionary.", file)
+        #     return
+
+        # print_or_write(f"--- Gap-filling statistics for {model_name} ---", file)
+        removed_reactions = self.reactions_removed[model_name]
+        num_removed_reactions = len(removed_reactions)
+        num_reconstructable = len(results["Reconstructable targets"])
+
+        reconstructable_reactions = set(removed_reactions) & set(results["Reconstructable targets"])
+        num_reconstructable_removed = len(reconstructable_reactions)
+
+        # print_or_write(f"Number of removed reactions: {num_removed_reactions}", file)
+        # print_or_write(f"Number of reconstructable targets: {num_reconstructable}", file)
+        # print_or_write(f"Number of removed reactions that are reconstructable: {num_reconstructable_removed}", file)
+        # print_or_write(f"Reconstructable reactions that were removed: {reconstructable_reactions}", file)
+        # print_or_write("----------------------------------------------------", file)
+
+    def save_results_to_file(self):
+        output_folder = 'meneco_outputs'
+        os.makedirs(output_folder, exist_ok=True)
+        output_file = os.path.join(output_folder, "all_models_results.txt")
+
+        with open(output_file, 'w') as f:
+            for model_name, results in self.results.items():
+                f.write(f"--- Meneco output for {model_name} ---\n")
+                f.write(f"Results for {model_name}:\n")
+                for key, value in results.items():
+                    f.write(f"{key}: {value}\n")
+
+                self.print_gapfilling_stats(model_name, results, file=f)
```
I don't quite understand why this test suite is in the examples instead of being located in the tests folder

jcapels commented on examples/random_knockout.py:
```diff
@@ -0,0 +1,68 @@
+    models_folder = '/Users/josediogomoura/gap_filling_dl/tests/data/original_model'
+    objective_function_id = 'Biomass_assembly_C3_in'
+    temporary_universal_model = True
+    performance_meneco_folders(models_folder, objective_function_id, temporary_universal_model)
```
I don't quite understand what this is doing... For what part of our experiments is this?

jcapels commented on service_main.py:
I believe the content herein presented should be split into parts and put in the README to document how to use our tool. This file can still be here, but I think the information should also be in the README file

jcapels commented on src/.DS_Store:
This file must be removed

jcapels commented on src/gap_filling_dl/.DS_Store:
same

jcapels commented on src/gap_filling_dl/COBRAmodel.py:
I don't quite understand why this file is here, as it is not called in any other class... this is not part of the package, right? this is only used to run only meneco without our workflow right? If so, it should be elsewhere, in my perspective

jcapels commented on src/gap_filling_dl/FilesUtilities.py:
is this related to the package? I think this is only for the worker and I thought you would put that in another repository or at least outside the source code... this should be a standalone project ready to deploy in Pypi...

jcapels commented on src/gap_filling_dl/MenecoRunner.py:
same

jcapels commented on src/gap_filling_dl/biomeneco/.DS_Store:
remove this file

jcapels commented on src/gap_filling_dl/biomeneco/fix_files/fix_terms.py:
```diff
@@ -0,0 +1,24 @@
+# fix_terms(path_seeds, path_seeds)
+#
+# path_targets = '/Users/josediogomoura/gap_filling_dl/tests/data/6gaps/6gaps_targets.xml'
+# fix_terms(path_targets, path_targets)
```
please do not leave unnecessary comments in the source code

jcapels commented on src/gap_filling_dl/biomeneco/fix_files/fix_underscore.py:
```diff
@@ -0,0 +1,19 @@
+
+# try with a file in a path:
+#fix_file("/Users/josediogomoura/gap_filling_dl/tests/performance_tests/data/kegg_universal_model.xml")
+# fix_file("/Users/josediogomoura/gap_filling_dl/tests/data/universal_model_kegg.xml")
```
same as above

jcapels commented on src/gap_filling_dl/biomeneco/gapfiller.py:
```diff
@@ -0,0 +1,1094 @@
+            bar.update(upd_value)
+
+
+parallelbar.parallelbar._process_status = _my_process_status
```
i think this could be a _utils file

jcapels commented on src/gap_filling_dl/biomeneco/gapfiller.py:
```diff
@@ -0,0 +1,1094 @@
+
+
+class GapFiller:
+    def __init__(self, model_path, universal_model_path, results_path, max_solutions):
```
this variables should be typed

jcapels commented on src/gap_filling_dl/biomeneco/gapfiller.py:
```diff
@@ -0,0 +1,1094 @@
+        ----------
+        model_path
+        universal_model_path
+        results_path
```
you should explain what each of the variables does... and write the type of the variable as well

jcapels commented on src/gap_filling_dl/my_classes.py:
this file is no longer used right? why is it here then?

jcapels commented on src/gap_filling_dl/worker.py:
this should be outside of source code... in another folder called web_service or something similar

jcapels commented on src/gap_filling_dl/write_sbml/metabolites_to_sbml.py:
why is this here? couldn't it be in the utils.py file?

jcapels commented on tests/unit_tests/fix_files/test_fix_terms.py:
```diff
@@ -0,0 +1,36 @@
+import unittest
+import os
+import tempfile
+from src.gap_filling_dl.biomeneco.fix_files.fix_terms import fix_terms
```
why src? this should be imported without the src?

jcapels commented on tests/unit_tests/fix_files/test_fix_underscore.py:
```diff
@@ -0,0 +1,28 @@
+import unittest
+import os
+from tempfile import TemporaryDirectory
+from src.gap_filling_dl.biomeneco.fix_files.fix_underscore import fix_file
```
same... you should leave the src out

jcapels reviewed (changes requested):
@beach-fossils, I came across several issues, mostly related to code cleanliness and organization.

I really think the following things should be improved before merging into the main branches and deploying to Pypi:

- There are different **Model** and **GapFiller** classes in the source code (one in the **model.py** and **gapfiller.py** and another in the **my_classes.py** files). I got quite confused with that and I still do not know which one is the one that you used to run the experiments. Furthermore, there are tests for both classes, which I found even more confusing.
- I really think you should create a README file documenting **(1)** how to run each step of the tool and **(2)** how to reproduce the results and check the code for each of the things you tested.
- There is one module that I was not able to import: ../src/gap_filling_dl/biomeneco/gapfiller.py:25: in <module>
    from write_sbml import write_metabolites_to_sbml
E   ModuleNotFoundError: No module named 'write_sbml'
-  The files related to the web service should be in another folder rather than in the source code folder.
- There are unnecessary comments all over the place, please remove the ones that do not document what you are doing in the code, exclusively.
- Last but not (definitely) least, please write the docstrings appropriately for all methods and classes, and make each variable type explicit both for the input and output variables of each method. Remove all .DS_store files from the repository.

After resolving these issues, I will review the code again and decide if it is ready for deployment.

beach-fossils committed:
[Update] ExampleSubmissionParameters.JSON

Atualizei isto para a tese com o novo param 'max_solutions'

beach-fossils committed:
Deleted classes directory

beach-fossils committed:
Remove .DS_Store files

beach-fossils committed:
Update README.md

Changed a bit the README file

beach-fossils committed:
Remove unused my_classes.py file

beach-fossils committed:
[ADD] Basic docstrings and created 'add_reactions_to_model_v2'

- Introduced 'add_reactions_to_model_v2', a more readable version of 'add_reactions_to_model'.
- Introduced several new helper methods to modularize the functionality:
  - 'get_mets_in_medium': Retrieves metabolites present in the model's medium.
  - 'get_completion': Extracts completion set from gap-filling results.
  - 'add_reactions_to_temp_model': Adds reactions to a temporary model and updates existing reactions.
  - 'create_sinks_for_metabolites': Creates sink reactions for metabolites in the model.
  - 'check_for_positive_solution': Checks for a positive solution in the temporary model.
  - 'add_demands_and_reoptimize': Adds demand reactions and reoptimizes the model.
  - 'handle_final_model_selection': Selects the final model based on available positive solutions or minimal completion.
  - 'handle_no_positive_solution': Handles cases with no positive solution by adding minimal completion to the model.
- The refactoring also includes adding detailed docstrings to improve code documentation and aid future development efforts.

beach-fossils committed:
[ADD] Comprehensive Unit Tests for GapFiller Class Methods

beach-fossils committed:
[UPDATE] Enhance code Readability and Documentation

Description:
- Added comprehensive docstrings to existing methods and classes to improve code documentation and understanding.
- Developed new methods aimed at enhancing code readability, ensuring easier maintenance and future development.
- Refactored some existing code segments for clarity, aligning with best practices for clean and readable code

beach-fossils committed:
[UPDATE] Unit Tests for Model Class in model.py

beach-fossils committed:
Delete examples/run_bioiso_and_meneco.py

beach-fossils committed:
Delete examples/random_knockout.py

beach-fossils committed:
Delete examples/data/6gaps directory

beach-fossils committed:
Delete examples/data/models directory

beach-fossils committed:
Delete examples/data directory

beach-fossils committed:
Delete tests/unit_tests/test_mini.py

beach-fossils committed:
Delete tests/data/models directory

beach-fossils committed:
Delete tests/data/test_toy directory

beach-fossils committed:
Delete tests/data/cvulgaris/output directory

beach-fossils committed:
Delete tests/data/6gaps directory

beach-fossils committed:
Delete tests/data/synechocystis/output directory

beach-fossils committed:
Delete tests/data/meneco/synechocystis/input directory

beach-fossils committed:
Delete tests/performance_tests/meneco_outputs directory

beach-fossils committed:
Delete tests/performance_tests/synechocystis.ipynb

beach-fossils committed:
Delete tests/performance_tests/minmal_completion_example.ipynb

beach-fossils committed:
Delete tests/performance_tests directory

beach-fossils committed:
Delete tests/meneco_outputs/all_models_results.txt

beach-fossils committed:
Delete tests/integration_tests directory

beach-fossils committed:
Delete src/gap_filling_dl/biomeneco/generate_targets.py

beach-fossils committed:
Delete src/gap_filling_dl/utils.py

beach-fossils committed:
Update utils.py

write_metabolites_to_sbml added to the file

beach-fossils committed:
Update meneco_only.py

beach-fossils committed:
Update gapfiller.py

beach-fossils committed:
Update model.py

beach-fossils committed:
Delete tests/unit_tests/write_sbml directory

beach-fossils committed:
Update test_fix_terms.py

beach-fossils committed:
Update test_fix_underscore.py

beach-fossils committed:
Delete src/gap_filling_dl/write_sbml directory

beach-fossils committed:
Relocated specific files from src to appropriate directories for better organization and clarity

beach-fossils committed:
Update fix_terms.py

beach-fossils committed:
Update fix_underscore.py

beach-fossils committed:
Update gapfiller.py

Added type annotations to constructor parameters

beach-fossils committed:
Moved the MennecoRunner.py

beach-fossils committed:
Update Dockerfile

beach-fossils committed:
Update Dockerfile

beach-fossils committed:
Update README.md

beach-fossils committed:
Update ci.yml

Included a step to run unit tests using for the 'biomeneco' module

beach-fossils committed:
Update test_gapfiller.py

beach-fossils committed:
Update test_gapfiller.py

beach-fossils committed:
Update ci.yml

beach-fossils committed:
Update utils.py

beach-fossils committed:
Update gapfiller.py
